### PR TITLE
feat: testing and configurable iterative mapping

### PR DESCRIPTION
**Pull request by tris203**
this refactors the mapping to take a config for custom maps, work on #16 

currently, it hits 5 of the example mapping functions

TODO: Hit the other mapping examples, if anyone has anymore please put them here or in the issue

Additionally, will need to keep the sensible defaults for the mapping definitions and offer a way to put them in through config and merge them with the defaults

### CODE REVIEW

**Review by willothy — APPROVED** [review]
lgtm once the queries are updated to use format strings

**Comment by tris203** [comment]
TODO:

 - Ability to set in config and merge in
 - Sensible defaults
 - Handle which key mappings - parse and hand off to ```which-key``` in a ```pcall```?
 - Handle lazy key maps ```(expression_list)```

File: lua/hawtkeys/ts.lua [hunk 1]
```diff
@@ -67,31 +67,46 @@ local scannedFiles = {}
...
+    ---@return string
+    local function buildArgs(params)
+        local args = ""
+        for name, opts in pairs(params) do
+            if opts.method == "dot_index_expression" then
+                args = args .. ' "' .. name .. '"'
+            end
         end
+        return args
     end
```
Comment (willothy) [inline lua/hawtkeys/ts.lua:86]:
```suggestion
```

Comment (willothy) [reply lua/hawtkeys/ts.lua:86]: oops accident

Comment (willothy) [inline lua/hawtkeys/ts.lua]:
```suggestion
```

File: lua/hawtkeys/ts.lua [hunk 2]
```diff
@@ -67,31 +67,46 @@ local scannedFiles = {}
 ---@param mapDefs keyMapArgs[]
```
Comment (willothy) [inline lua/hawtkeys/ts.lua]:
```suggestion
---@param params keyMapArgs[]
---@param method string
---@return string
local function buildArgs(params, method)
    local args = ""
    for name, opts in pairs(params) do
        if opts.method == method then
            args = args .. ' "' .. name .. '"'
        end
    end
    return args
end

---@param mapDefs keyMapArgs[]
```

File: lua/hawtkeys/ts.lua [hunk 3]
```diff
@@ -67,31 +67,46 @@ local scannedFiles = {}
...
+    ---@return string
+    local function buildArgs(params)
+        local args = ""
+        for name, opts in pairs(params) do
+            if opts.method == "function_call" then
+                args = args .. ' "' .. name .. '"'
+            end
         end
+        return args
     end
```
Comment (willothy) [inline lua/hawtkeys/ts.lua]:
```suggestion
```

File: lua/hawtkeys/ts.lua [hunk 4]
```diff
@@ -67,31 +67,46 @@ local scannedFiles = {}
...
-    query = query .. ")"
-    query = query .. "\n(arguments) @args)"
-    return query
+    return string.format(query, buildArgs(mapDefs))
```
Comment (willothy) [inline lua/hawtkeys/ts.lua]:
```suggestion
    return string.format(query, buildArgs(mapDefs, "function_call"))
```

File: lua/hawtkeys/ts.lua [hunk 5]
```diff
@@ -67,31 +67,46 @@ local scannedFiles = {}
...
+            if opts.method == "dot_index_expression" then
+                args = args .. ' "' .. name .. '"'
+            end
         end
+        return args
     end
-    query = query .. ")"
-    query = query .. "\n(arguments) @args)"
-    return query
+    return string.format(query, buildArgs(mapDefs))
```
Comment (willothy) [inline lua/hawtkeys/ts.lua]:
```suggestion
    return string.format(query, buildArgs(mapDefs, "dot_index_expression"))
```
This way we avoid creating closures for each call, and reuse the buildArgs function

**Review by willothy — CHANGES_REQUESTED** [review]
should reuse the buildArgs function imo

File: lua/hawtkeys/ts.lua [hunk 6]
```diff
@@ -6,7 +6,132 @@ local config = require("hawtkeys")
...
+    }, -- for my personal config - used in lsp-setup
+}
+
+---@type table<string, boolean>
 local scannedFiles = {}
+
+---@param params keyMapArgs[]
+---@param method setMethods
+---@return string
+local function buildArgs(params, method)
```
Comment (willothy) [inline lua/hawtkeys/ts.lua]: maybe `build_args` would be better for consistency?

**Review by willothy — APPROVED** [review]

**Comment by tris203** [comment]
also adds frame work for testing needed for #17

File: lua/hawtkeys/init.lua [hunk 7]
```diff
@@ -1,25 +1,69 @@
...
+---@field leader string | nil
+---@field homerow number | nil
+---@field powerFingers number[] | nil
+---@field keyboardLayout SupportedKeyboardLayouts | nil
+---@field customMaps { [string] : TSKeyMapArgs | WhichKeyMapargs } | nil
+---
+
+---@type { [string] : TSKeyMapArgs | WhichKeyMapargs }---
+
+M._defaultSet = {
```
Comment (willothy) [inline lua/hawtkeys/init.lua]: Should this be a local? Seems like it shouldn't need to be exposed externally other than through config

File: lua/hawtkeys/init.lua [hunk 8]
```diff
@@ -1,25 +1,69 @@
...
+    M.keyMapSet =
+        vim.tbl_extend("force", M._defaultSet, config.customMaps or {})
 end
 
+vim.api.nvim_create_user_command(
+    "Hawtkeys",
+    "lua require('hawtkeys.ui').show()",
+    {}
+)
+vim.api.nvim_create_user_command(
```
Comment (willothy) [inline lua/hawtkeys/init.lua]: I think these should go in `setup`

**Review by willothy — APPROVED** [review]
lgtm!

**Pull request merged**